### PR TITLE
A potentially better way to fine-tune BERT for contextual tasks

### PR DESCRIPTION
**Pull request by s-jse**
- Using `--append_question_to_context_too` will append a copy of the question to the end of context, and will tokenize it like a non-program field.
This approach seems effective especially for the `almond_dialogue_nlu` task.
- transformers is updated to version 2.9. Due to changes in model hashes, rebuilding the common docker image might be necessary.

### CODE REVIEW

File: genienlp/arguments.py [hunk 1]
```diff
@@ -175,6 +175,8 @@ def parse_argv(parser):
                         help='probability of replacing a token with mask for MLM pretraining')
     parser.add_argument('--force_subword_tokenize', action='store_true', default=False,
                         help='force subword tokenization of code tokens too')
+    parser.add_argument('--append_question_to_context_too', action='store_true', default=False,
```
Comment (gcampax) [inline genienlp/arguments.py:178]: The argument name is tad verbose, IMHO

File: genienlp/data_utils/example.py [hunk 2]
```diff
@@ -45,20 +45,23 @@ class Example(NamedTuple):
...
+    context_plus_question_word_mask: List[bool]
 
     vocab_fields = ['context', 'question', 'answer']
 
     @staticmethod
     def from_raw(example_id: str, context: str, question: str, answer: str, tokenize, lower=False):
         args = [example_id]
-        for argname, arg in (('context', context), ('question', question), ('answer', answer)):
+        for argname, arg in (('context', context), ('question', question), ('answer', answer), ('question', context+' '+question)):
             words, mask = tokenize(arg.rstrip('\n'), field_name=argname)
```
Comment (gcampax) [inline genienlp/data_utils/example.py:57]: I see we pass 'question' here to the Task class, so for Almond tasks everything will be subword split (mask all True). Just confirming that's the expected behavior.

File: genienlp/tasks/almond/__init__.py [hunk 3]
```diff
@@ -233,7 +234,10 @@ class AlmondDialogueNLU(BaseAlmondTask):
     state of the conversation)
     """
     def _is_program_field(self, field_name):
-        return field_name in ('answer', 'context')
+        if self.append_question_to_context_too:
+            return field_name == 'answer'
+        else:
+            return field_name in ('answer', 'context')
```
Comment (gcampax) [inline genienlp/tasks/almond/__init__.py]: You don't need this, cause you'll get "question" as the field name for the concatenated field?

**Review by gcampax — APPROVED** [review]
This looks good. Clean and neat. Looking forward to try it.
Couple comments just to check.

**Comment by gcampax** [comment]
Btw, for the transformers version you need to change setup.py too, because we don't use pipenv in the docker.
(In fact, we should probably get rid of the Pipfile entirely, it's just confusing)

File: genienlp/tasks/almond/__init__.py (hunk 3 above)
Comment (s-jse) [reply genienlp/tasks/almond/__init__.py]: Yes, it is a remnant of another approach. I'll remove it.

File: genienlp/data_utils/example.py (hunk 2 above)
Comment (s-jse) [reply genienlp/data_utils/example.py:57]: Yes, masks will be all True except for words that are entities (start with capital letters).

File: genienlp/arguments.py (hunk 1 above)
Comment (s-jse) [reply genienlp/arguments.py:178]: This option appends the question to context, but feeds in the original question separately as well, hence "too". What do you suggest?

Comment (gcampax) [reply genienlp/arguments.py:178]: I was going to suggest just "append_question_to_context". Shorter and equally unambiguous.

File: genienlp/run_generation.py [hunk 4]
```diff
@@ -152,8 +152,8 @@ def sample_sequence(model, length, context, position_ids, num_samples=1, tempera
                 inputs['past'] = past
                 if past is not None:
                     inputs['input_ids'] = next_token
-                    inputs['position_ids'] = position_ids[:, next_index-1]
-                    inputs['token_type_ids'] = segment_ids[:, next_index-1]
+                    inputs['position_ids'] = position_ids[:, next_index-1].unsqueeze(-1)
+                    inputs['token_type_ids'] = segment_ids[:, next_index-1].unsqueeze(-1)
```
Comment (gcampax) [inline genienlp/run_generation.py:156]: This was not "Remove unused code". Sneaked in by accident?

Comment (s-jse) [reply genienlp/run_generation.py:156]: Should have been two separate commits, but the change was intentional. It failed one of the paraphrasing tests with the newer transformers version, so I had to make this change. All this code will be obsolete when we merge `wip/paraphrase` branch.

**Pull request merged**